docs/proposals/remote-artifacts.md

```diff
@@ -0,0 +1,293 @@
+following example:
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: Build
+metadata:
+  name: license-file
+spec:
+  sources:
+    - name: license-file
+      type: http
+      url: https://licenses.company.com/customer-id/license.tar
+      http:
+        path: $(workspace)/licenses/license.tar
+```
+
+The `source` section will be renamed to `sources` in order to accommodate more `type`s of inputs.
+Each entry will be composed of the following attributes:
+
+- `name`: source name (required);
+- `type`: input source type, `git` or `http` (required);
+- `sourceRef`: use a external resource, the name in this field is the resource name (optional);
+- `url`: the URL of the repository or remote artifact (optional);
+- `credentials`: the credentials to interact with the artifact (optional);
+- `http`: settings for `http`, namely `path` (optional);
+- `git`: settings for `git`, namely `revision` (optional);
+
+The resource `UID`/`GID` will be defined by the same user running the Tekton task, later on, we
+can extend the API to support arbitrary configuration.
+
+### Standalone CRD
+
+Alternatively, we may define the artifacts as a standalone CRD, that is a `BuildSource`
+resource. The advantage of this design is being able to exchange, and reuse, artifacts on several
+builds. For instance, if two projects are sharing a common logo image, both `Builds` will refer to
+the same `BuildSource`.
+
+The following snippet shows how an Artifact (`BuildSource`) will be represented.
+
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: BuildSource
+metadata:
+  name: license-file
+spec:
+  sources:
+    - type: http
+      url: https://licenses.company.com/customer-id/license.tar
+      http:
+        path: $(workspace)/licenses/license.tar
+```
+
+## Usage
+
+The usage of this feature is based on declaring a slice of `.spec.sources` and later on overwriting
+entries using `BuildRun` resource. For instance:
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: Build
+metadata:
+  name: example
+spec:
+  sources:
+    - name: license-file
+```
+
+Could have its `license-file` overwritten in a `BuildRun` with:
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: BuildRun
+metadata:
+  name: license-file
+spec:
+  buildRef:
+    name: example
+  sources:
+    - name: license-file
+      sourceRef: alternative-file
+```
+
+## Steps and Helper
+
+Artifacts are requirements for the build process, thus downloading and preparing artifacts must
+happen before the build process starts. To achieve the objective, the operator will generate a new
+task step to download the artifacts.
+
+The download may happen using existing open-source software, more specifically `wget`. The
+container image which contains this software will be specified via environment variable, and when
+not informed will use a default value.
+
+Therefore, the operator needs to generate `wget` commands with arguments in order to download the
```

Comment:
> the operator needs to generate `wget` commands with arguments in order to download the

_The buildrun controller needs to generate a TaskRun step that would generate the `wget` command to download the artifacts specified_

Reply:
Thanks, that's a good addition. Please consider latest commit (`01ace39`)